cpmpy/tools/explain/mus.py

```diff
@@ -184,3 +236,85 @@ def do_recursion(soft, hard, delta):
 
     core = do_recursion(soft, hard, [])
     return core
+
+def optimal_mus_naive(soft, hard=[], weights=1, solver="ortools", hs_solver="ortools", grow=mss_grow_naive):
+    """
+        Naive implementation of `optimal_mus` without assumption variables and incremental solving
+    """
+
+    soft = toplevel_list(soft, merge_and=False)
+    bvs = cp.boolvar(shape=len(soft))
+
+    if is_num(weights):
+        weights = np.ones(len(bvs), dtype=int)
+    hs_solver = cp.SolverLookup.get(hs_solver)
+    hs_solver.minimize(cp.sum(bvs * np.array(weights)))
+
+    while hs_solver.solve() is True:
+
+        hitting_set = [c for bv, c in zip(bvs, soft) if bv.value()]
+        if cp.Model(hard + hitting_set).solve(solver=solver) is False:
+            break
+        # SAT, how to do grow without assumptions?
+        #   for now, using naive grow methods from tools/explain/mcs
+        mss = grow(soft=[c for bv, c in zip(bvs, soft) if not bv.value()],
+                   hard=hitting_set + hard,
+                   solver=solver)
+        mcs = [bv for bv,c in zip(bvs, soft) if c not in frozenset(hitting_set + mss)]
+        hs_solver += cp.sum(mcs) >= 1
+
+    return hitting_set
+
+
+## MUS enumeration (should this be in a separate file?)
+def marco(soft, hard=[], solver="ortools", map_solver="ortools"):
+    """
+        Enumerates all MUSes in the unsatisfiable problem
+        Iteratively generates a subset of constraints (the seed) and checks whether it is SAT.
+        When the seed is SAT, it is grown to an MSS to derive an MCS.
+        This MCS is then added to the Map solver as a set to hit
+        In case the seed is UNSAT, the seed is shrunk to a real MUS and returned.
+        The Map-solver is instructed to not generated any superset of this MUS as next seeds
+
+        Based on:
+            Liffiton, Mark H., et al. "Fast, flexible MUS enumeration." Constraints 21 (2016): 223-250.
+    """
+
+    model, soft, assump = make_assump_model(soft, hard)
+    dmap = dict(zip(assump, soft))
+    s = cp.SolverLookup.get(solver, model)
+
+    map_solver = cp.SolverLookup.get(map_solver)
+    map_solver += cp.any(assump)
+    map_solver.solution_hint(assump, [1]*len(assump)) # we want large subsets, more likely to be a MUS
```

Comment:
careful, I think this is ortools specific... maybe check if interface has this function